parsers/keyval/keyval.go

```diff
@@ -0,0 +1,216 @@
+	nower      Nower
+
+	warnedAboutTime bool
+}
+
+type Nower interface {
+	Now() time.Time
+}
+
+type RealNower struct{}
+
+func (r *RealNower) Now() time.Time {
+	return time.Now().UTC()
+}
+
+func (p *Parser) Init(options interface{}) error {
+	p.conf = *options.(*Options)
+
+	p.nower = &RealNower{}
+	p.lineParser = &KeyValLineParser{}
+	return nil
+}
+
+type LineParser interface {
+	ParseLine(line string) (map[string]interface{}, error)
+}
+
+type KeyValLineParser struct {
+}
+
+func (j *KeyValLineParser) ParseLine(line string) (map[string]interface{}, error) {
+	parsed := make(map[string]interface{})
+	f := func(key, val []byte) error {
+		keyStr := string(key)
+		valStr := string(val)
+		if b, err := strconv.ParseBool(valStr); err == nil {
+			parsed[keyStr] = b
+			return nil
+		}
+		if i, err := strconv.Atoi(valStr); err == nil {
+			parsed[keyStr] = i
+			return nil
+		}
+		if f, err := strconv.ParseFloat(valStr, 64); err == nil {
+			parsed[keyStr] = f
+			return nil
+		}
+		parsed[keyStr] = valStr
+		return nil
+	}
+	err := logfmt.Unmarshal([]byte(line), logfmt.HandlerFunc(f))
+	return parsed, err
+}
+
+func (p *Parser) ProcessLines(lines <-chan string, send chan<- event.Event, prefixRegex *parsers.ExtRegexp) {
+	for line := range lines {
+		logrus.WithFields(logrus.Fields{
+			"line": line,
+		}).Debug("Attempting to process json log line")
+
+		// take care of any headers on the line
+		var prefixFields map[string]string
+		if prefixRegex != nil {
+			var prefix string
+			prefix, prefixFields = prefixRegex.FindStringSubmatchMap(line)
+			line = strings.TrimPrefix(line, prefix)
+		}
+
+		parsedLine, err := p.lineParser.ParseLine(line)
+
+		if err != nil {
+			// skip lines that won't parse
+			logrus.WithFields(logrus.Fields{
+				"line": line,
+			}).Debug("skipping line; failed to parse.")
+			continue
+		}
+		// merge the prefix fields and the parsed line contents
+		for k, v := range prefixFields {
+			parsedLine[k] = v
+		}
+
+		// look for the timestamp in any of the prefix fields or regular content
+		timestamp := p.getTimestamp(parsedLine)
+
+		// send an event to Transmission
+		e := event.Event{
+			Timestamp: timestamp,
+			Data:      parsedLine,
+		}
+		send <- e
+	}
+	logrus.Debug("lines channel is closed, ending json processor")
+}
+
+// getTimestamp looks through the event map for something that looks
+// like a timestamp. It will guess at the key name or use
+// the one from Config if it is not ""
+// if unable to parse it will return the current time
+// it is highly recommended that you populate the Config with time format
+// sample from logrus: "time":"2014-03-10 19:57:38.562264131 -0400 EDT"
+func (p *Parser) getTimestamp(m map[string]interface{}) time.Time {
+	var ts time.Time
+	if p.conf.TimeFieldName != "" {
+		// remove the timestamp from the body when we stuff it in the header
+		defer delete(m, p.conf.TimeFieldName)
+		if t, found := m[p.conf.TimeFieldName]; found {
+			ts = p.tryTimeFormats(t.(string))
```

Comment:
Is it definitely safe to cast t to a string at this point? What if the logfile has something like `timestamp=1490664217`?